packages/grid/_modules_/grid/hooks/features/rows/useEditRows.ts

```diff
@@ -0,0 +1,154 @@
+import { GridState } from '../core/gridState';
+import { useGridSelector } from '../core/useGridSelector';
+import { useGridState } from '../core/useGridState';
+
+export interface EditCellProps {
+  value: CellValue;
+  [prop: string]: any;
+}
+
+export type EditRow = { [field: string]: true | EditCellProps };
+export type EditRowsModel = { [rowId: string]: EditRow };
+
+export const editRowsStateSelector = (state: GridState) => state.editRows;
+
+export function useEditRows(apiRef: ApiRef) {
+  const [, setGridState, forceUpdate] = useGridState(apiRef);
+  const options = useGridSelector(apiRef, optionsSelector);
+
+  const setCellEditMode = React.useCallback(
+    (id, field) => {
+      setGridState((state) => {
+        const currentCellEditState: EditRowsModel = { ...state.editRows };
+        currentCellEditState[id] = currentCellEditState[id] || {};
+        currentCellEditState[id][field] = true;
+
+        const newEditRowsState = { ...state.editRows, ...currentCellEditState };
+
+        apiRef.current.publishEvent(EDIT_ROW_MODEL_CHANGE, newEditRowsState);
+        return { ...state, editRows: newEditRowsState };
+      });
+      apiRef.current.publishEvent(CELL_MODE_CHANGE, {
+        id,
+        field,
+        mode: 'edit',
+        api: apiRef.current,
+      });
+      forceUpdate();
+    },
+    [apiRef, forceUpdate, setGridState],
+  );
+
+  const setCellViewMode = React.useCallback(
+    (id, field) => {
+      setGridState((state) => {
+        const newEditRowsState: EditRowsModel = { ...state.editRows };
+
+        if (!newEditRowsState[id]) {
+          return state;
+        }
+
+        if (newEditRowsState[id][field]) {
+          delete newEditRowsState[id][field];
+          if (!Object.keys(newEditRowsState[id]).length) {
+            delete newEditRowsState[id];
+          }
+        }
+        apiRef.current.publishEvent(EDIT_ROW_MODEL_CHANGE, newEditRowsState);
+
+        return { ...state, editRows: newEditRowsState };
+      });
+      apiRef.current.publishEvent(CELL_MODE_CHANGE, {
+        id,
+        field,
+        mode: 'view',
+        api: apiRef.current,
+      });
+      forceUpdate();
+    },
+    [apiRef, forceUpdate, setGridState],
+  );
+
+  const setCellMode = React.useCallback(
+    (id, field, mode: CellMode) => {
+      if (mode === 'edit') {
+        setCellEditMode(id, field);
+      } else {
+        setCellViewMode(id, field);
+      }
+    },
+    [setCellEditMode, setCellViewMode],
+  );
+
+  const isCellEditable = React.useCallback(
+    (params: CellParams) => {
+      return params.colDef.editable && (!options.isCellEditable || options.isCellEditable(params));
+    },
+    [options.isCellEditable],
+  );
+
+  const commitCellValueChanges = React.useCallback(
+    (update: RowModelUpdate) => {
+      if (apiRef.current.hasListener(CELL_VALUE_CHANGE_COMMITTED)) {
+        apiRef.current.publishEvent(CELL_VALUE_CHANGE_COMMITTED, { update, api: apiRef.current });
+        return;
```

Comment:
We can't change the behavior because a listener is added or not. I don't think it's negotiable.

```suggestion

```

Reply:
Yes I will refactor that. I was surprised there was no some kind of `hasListener` method, it depends of the `eventEmitter` implementation